pkg/server/nodemanager/node_controller.go

```diff
@@ -267,13 +270,45 @@ func (c *NodeController) startSingleNode(node *api.Node, image cloud.Image, clou
 	var (
 		instanceID string
 		err        error
+		excludeInstances []string
 	)
-	if node.Spec.Spot {
-		instanceID, err = c.CloudClient.StartSpotNode(node, image, cloudInitData, c.Config.DefaultIAMPermissions)
-	} else if node.Spec.Dedicated {
-		instanceID, err = c.CloudClient.StartDedicatedNode(node, image, cloudInitData, c.Config.DefaultIAMPermissions)
-	} else {
-		instanceID, err = c.CloudClient.StartNode(node, image, cloudInitData, c.Config.DefaultIAMPermissions)
+	for retries := 0; retries < NoCapacityRetryLimit; retries++ {
+		if node.Spec.Spot {
+			instanceID, err = c.CloudClient.StartSpotNode(node, image, cloudInitData, c.Config.DefaultIAMPermissions)
+		} else if node.Spec.Dedicated {
+			instanceID, err = c.CloudClient.StartDedicatedNode(node, image, cloudInitData, c.Config.DefaultIAMPermissions)
+		} else {
+			instanceID, err = c.CloudClient.StartNode(node, image, cloudInitData, c.Config.DefaultIAMPermissions)
+		}
+		if err != nil && strings.Contains(err.Error(), "InsufficientInstanceCapacity") {
+			// detect Region has no capacity: InsufficientInstanceCapacity
+			// and try to find matching instance other than the one with no capacity in region,
+			// then, try again to start a node.
+			// We collect instance type which failed with InsufficientInstanceCapacity to excludeInstances slice,
+			// to avoid loop:
+			// choose instance type ->
+			//   instance A chosen ->
+			//     booting instance A failed ->
+			//       choose instance type ->
+			//         instance B chosen ->
+			//           booting instance B failed ->
+			//             instance A chosen ->
+			//                etc.
+			// for safety reasons, we also set maximum number of retries.
+			excludeInstances = append(excludeInstances, node.Spec.InstanceType)
+			instanceType, _ := instanceselector.GetInstanceFromResources(node.Spec.Resources, func(inst instanceselector.InstanceData) bool {
+				for _, excludedInstanceType := range excludeInstances {
+					if inst.InstanceType == excludedInstanceType {
+						return false
+					}
+				}
+				return true
+			})
+			klog.V(2).Infof("No sufficient capacity in region to boot %s , trying to boot %s", node.Spec.InstanceType, instanceType)
+			node.Spec.InstanceType = instanceType
+		} else {
+			break
+		}
```

Comment:
We already have `pkg/server/nodemanager/boot_limiter.go` for keeping track of unavailable instances. See https://github.com/elotl/kip/blob/master/pkg/server/nodemanager/node_scaler.go#L85-L111 for a check where we're already using it to limit boot requests for unavailable instance types.